lld/COFF/Symbols.cpp

```diff
@@ -126,9 +127,14 @@ DefinedImportThunk::DefinedImportThunk(COFFLinkerContext &ctx, StringRef name,
 
 Defined *Undefined::getWeakAlias() {
   // A weak alias may be a weak alias to another symbol, so check recursively.
-  for (Symbol *a = weakAlias; a; a = cast<Undefined>(a)->weakAlias)
+  SmallSet<Symbol *, 4> weakChain;
+  for (Symbol *a = weakAlias; a; a = cast<Undefined>(a)->weakAlias) {
+    if (weakChain.contains(a))
+      break; // We have a cycle.
     if (auto *d = dyn_cast<Defined>(a))
       return d;
+    weakChain.insert(a);
```

Comment:
driveby comment: Normally it is faster to do something like

    if (!weakChain.insert(a).second)
        break; // we have a cycle

i.e. do the lookup&insert in one operation

Reply:
Good point, I changed that in the new version (used `try_emplace` after changing set type to `DenseMap`).